In [41]:
import numpy as np
import pandas as pd
from pathlib import Path
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import classification_report
from sklearn.ensemble import RandomForestClassifier
from sklearn.preprocessing import StandardScaler, MinMaxScaler, LabelEncoder

In [25]:
train_df = pd.read_csv(Path('Resources/2019loans.csv'))
test_df = pd.read_csv(Path('Resources/2020Q1loans.csv'))

In [26]:
train_df

loan_amnt  int_rate  installment home_ownership  annual_inc  \
0         7000.0    0.1894       256.38       MORTGAGE     75000.0   
1        40000.0    0.1614       975.71       MORTGAGE    102000.0   
2        11000.0    0.2055       294.81           RENT     45000.0   
3         4000.0    0.1612       140.87       MORTGAGE     38000.0   
4        14000.0    0.1797       505.93       MORTGAGE     43000.0   
...          ...       ...          ...            ...         ...   
12175    19975.0    0.2565       801.09           RENT     28000.0   
12176    15000.0    0.1774       540.34           RENT     50000.0   
12177     3600.0    0.1862       131.28           RENT     60000.0   
12178    15000.0    0.0881       475.68       MORTGAGE     62000.0   
12179    15000.0    0.1774       540.34           RENT     54080.0   

      verification_status pymnt_plan    dti  delinq_2yrs  inq_last_6mths  ...  \
0            Not Verified          n  28.62          0.0             2.0  ...   
1         Source Verified          n  11.72          2.0             0.0  ...   
2                Verified          n  37.25          1.0             3.0  ...   
3            Not Verified          n  42.89          1.0             0.0  ...   
4         Source Verified          n  22.16          1.0             0.0  ...   
...                   ...        ...    ...          ...             ...  ...   
12175        Not Verified          n  28.42          0.0             0.0  ...   
12176            Verified          n  23.43          4.0             0.0  ...   
12177        Not Verified          n  28.80          0.0             1.0  ...   
12178     Source Verified          n  11.44          0.0             0.0  ...   
12179        Not Verified          n  21.66          1.0             1.0  ...   

       percent_bc_gt_75  pub_rec_bankruptcies  tax_liens  tot_hi_cred_lim  \
0                  87.5                   0.0        0.0         352260.0   
1                   0.0                   0.0        0.0         294664.0   
2                   7.7                   0.0        0.0          92228.0   
3                 100.0                   0.0        0.0         284273.0   
4                  25.0                   0.0        0.0         120280.0   
...                 ...                   ...        ...              ...   
12175              16.7                   0.0        0.0          50055.0   
12176              11.1                   0.0        0.0          70324.0   
12177               0.0                   0.0        0.0          83765.0   
12178               0.0                   0.0        0.0         189930.0   
12179             100.0                   0.0        0.0          48968.0   

      total_bal_ex_mort  total_bc_limit  total_il_high_credit_limit  \
0               62666.0         35000.0                     10000.0   
1              109911.0          9000.0                     71044.0   
2               36007.0         33000.0                     46328.0   
3               52236.0         13500.0                     52017.0   
4               88147.0         33300.0                     78680.0   
...                 ...             ...                         ...   
12175           28192.0         18700.0                     19055.0   
12176           57025.0         13300.0                     54824.0   
12177           55156.0         14800.0                     53065.0   
12178           23748.0          7000.0                     32930.0   
12179           29786.0          2900.0                     33968.0   

       hardship_flag  debt_settlement_flag     target  
0                  N                     N   low_risk  
1                  N                     N   low_risk  
2                  N                     N   low_risk  
3                  N                     N   low_risk  
4                  N                     N   low_risk  
...              ...                   ...        ...  
12175              N   

In [49]:
# Convert categorical data to numeric and separate target feature for training data
X_train = train_df.drop("target", axis=1)
y_train = train_df["target"].values
y_train_label = LabelEncoder().fit_transform(y_train)

In [54]:
# Convert categorical data to numeric and separate target feature for testing data
X_test = test_df.drop("target",axis=1)
y_test = test_df["target"].values
y_test_label=LabelEncoder().fit_transform(y_test)

92

In [57]:
# add missing dummy variables to testing set
X_train_dummies = pd.get_dummies(X_train)
X_test_dummies = pd.get_dummies(X_test)
# print(len(X_train_dummies.columns))
# print(len(X_test_dummies.columns))

for i in X_train_dummies.columns:
    if i not in X_test_dummies.columns:
        X_test_dummies[i]=0


In [58]:
print(len(X_train_dummies.columns))
print(len(X_test_dummies.columns))


92
92


In [59]:
# Train the Logistic Regression model on the unscaled data and print the model score
classifier = LogisticRegression()
classifier.fit(X_train_dummies, y_train_label)
print(f"Training Data Score: {classifier.score(X_train_dummies, y_train_label)}")
print(f"Testing Data Score: {classifier.score(X_test_dummies, y_test_label)}")

Training Data Score: 0.6536124794745485
Testing Data Score: 0.5104210974053595


c:\Users\foh01\anaconda3\lib\site-packages\sklearn\linear_model\_logistic.py:763: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  n_iter_i = _check_optimize_result(


In [61]:
# Train a Random Forest Classifier model and print the model score
clf = RandomForestClassifier(random_state=1, n_estimators=500).fit(X_train_dummies, y_train_label)
print(f'Training Score: {clf.score(X_train_dummies, y_train_label)}')
print(f'Testing Score: {clf.score(X_test_dummies, y_test_label)}')

Training Score: 1.0
Testing Score: 0.646958740961293


In [62]:
# Scale the data
scaler = StandardScaler().fit(X_train_dummies)
X_train_scaled = scaler.transform(X_train_dummies)
X_test_scaled = scaler.transform(X_test_dummies)

In [63]:
# Train the Logistic Regression model on the scaled data and print the model score
classifier = LogisticRegression()
classifier.fit(X_train_scaled, y_train_label)
print(f"Training Data Score: {classifier.score(X_train_scaled, y_train_label)}")
print(f"Testing Data Score: {classifier.score(X_test_scaled, y_test_label)}")

Training Data Score: 0.710919540229885
Testing Data Score: 0.7598894087622289


c:\Users\foh01\anaconda3\lib\site-packages\sklearn\linear_model\_logistic.py:763: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  n_iter_i = _check_optimize_result(


In [64]:
# Train a Random Forest Classifier model on the scaled data and print the model score
clf = RandomForestClassifier(random_state=1, n_estimators=500).fit(X_train_scaled, y_train_label)
print(f'Training Score: {clf.score(X_train_scaled, y_train_label)}')
print(f'Testing Score: {clf.score(X_test_scaled, y_test_label)}')

Training Score: 1.0
Testing Score: 0.6480221182475542


Logistic Regression improved from unscaled 0.51 to scaled 0.76 while Random Forest went from 0.647 to 0.648.